### PR TITLE
[2567] Improve the API of sirius-web-application and the property section contributions

### LINKED ISSUE #2567: Fix the inconsistency in the typing of the custom widget API
In the API of the form property section, some parts of the API are relying on generics but they cannot possibly work. They appear to work because they have not been applied to existing property sections and the only place where they have been applied have very lax typeScript settings and thus pretty much anything would appear to "work". The issue comes from interfaces introduced for the support of custom widgets like `PropertySectionComponent` and `PropertySectionComponentProps`. They look nice with their generic signature but they do not really work. Here are the interfaces in question:

```
export type PropertySectionComponentProps<W extends GQLWidget> = {
  editingContextId: string;
  formId: string;
  widget: W;
  subscribers: GQLSubscriber[];
  readOnly: boolean;
  setSelection: (selection: Selection) => void;
};

export type PropertySectionComponent<W extends GQLWidget> = (
  props: PropertySectionComponentProps<W>
) => JSX.Element | null;
```

In `PropertySection.tsx`, we are not using generics at all because nothing in the existing API is based on generics so the type of `widget` is `GQLWidget`:

```
const CustomWidgetComponent = propertySectionsRegistry.getComponent(widget);
```

The property section registry is not leveraging generics in its function `getComponent` at all, which is the start of our problems:

```
export type PropertySectionComponentRegistry = {
  getComponent: (widget: GQLWidget) => PropertySectionComponent<GQLWidget> | null;
};
```

Which brings us to the code which reveals the issue:

```
const propertySectionsRegistry: PropertySectionComponentRegistry = {
  getComponent: (widget: GQLWidget) => {
    if (widget.__typename === 'Slider') {
      return SliderPropertySection;
    } else if (widget.__typename === 'ReferenceWidget') {
      return ReferencePropertySection;
    }
  }
}
```

Here, we can return a `SliderPropertySection` in case of a `SQLSlider` but the `SliderPropertySection` has to follow the signature of `PropertySectionComponent<W extends GQLWidget>`. As a result, its props have the type `PropertySectionComponentProps<GQLSlider>` which indicates that it can only work if it receives a `GQLSlider` but the signature of `PropertySectionComponentRegistry#getComponent` requires a `PropertySectionComponent<GQLWidget>` which would indicate that every single widget contributed using this API should have a props typed with `PropertySectionComponent<GQLWidget>` and not `PropertySectionComponent<W extends GQLWidget>`.

One could say that the fix is quite easy, the registry should be updated to have the following signature:

```
export type PropertySectionComponentRegistry = {
  getComponent: <Widget extends GQLWidget>(widget: Widget) => PropertySectionComponent<Widget> | null;
}
```

Unfortunately this does not work since we stumble on this error:

```
Type 'PropertySectionComponentProps<Widget>' is not assignable to type 'PropertySectionComponentProps<GQLSlider>'.
      Type 'Widget' is not assignable to type 'GQLSlider'.
        Type 'GQLWidget' is missing the following properties from type 'GQLSlider': minValue, maxValue, currentValue
```

And there starts a great adventure fighting TypeScript typing system with generics...

### PR DESCRIPTION
**Pull request by sbegaudeau**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/2985

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/forms/frontend/sirius-components-forms/src/form/FormEventFragments.ts [hunk 1]
```diff
@@ -314,6 +314,14 @@ export const widgetFields = (contributions: Array<WidgetContribution>) => `
...
   }
 
+  fragment sliderFields on Slider {
+    label
+    iconURL
+    minValue
+    maxValue
+    currentValue
+  }
+
```
Comment (pcdavid) [inline packages/forms/frontend/sirius-components-forms/src/form/FormEventFragments.ts:324]: Isn't the `... on Slider { ...sliderFields }` usage in `widgetFields` missing?

File: packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormEditPlugin.java [hunk 2]
```diff
@@ -90,15 +90,8 @@ protected static String extensionFor(String key) {
...
-                "gif".equalsIgnoreCase(extension) ||
-                "bmp".equalsIgnoreCase(extension) ||
-                "ico".equalsIgnoreCase(extension) ||
-                "jpg".equalsIgnoreCase(extension) ||
-                "jpeg".equalsIgnoreCase(extension) ||
-                "tif".equalsIgnoreCase(extension) ||
-                "tiff".equalsIgnoreCase(extension) ||
-                "svg".equalsIgnoreCase(extension)) {
+            if ("png".equalsIgnoreCase(extension) || "gif".equalsIgnoreCase(extension) || "bmp".equalsIgnoreCase(extension) || "ico".equalsIgnoreCase(extension) || "jpg".equalsIgnoreCase(extension)
+                    || "jpeg".equalsIgnoreCase(extension) || "tif".equalsIgnoreCase(extension) || "tiff".equalsIgnoreCase(extension) || "svg".equalsIgnoreCase(extension)) {
```
Comment (pcdavid) [inline packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormEditPlugin.java]: Is this reformatting on purpose?
We could also have something like:
```java
private static final List<String> IMAGE_EXTENSIONS = List.of(...);
...
return IMAGE_EXTENSIONS.stream().anyMatch(supportedExtension -> supportedExtension.equalsIgnoreCase(extension));
```

**Review by pcdavid — CHANGES_REQUESTED** [review]

File: packages/forms/frontend/sirius-components-forms/src/form/FormEventFragments.ts (hunk 1 above)
Comment (sbegaudeau) [reply packages/forms/frontend/sirius-components-forms/src/form/FormEventFragments.ts:324]: Done

File: packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormEditPlugin.java (hunk 2 above)
Comment (sbegaudeau) [reply packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormEditPlugin.java]: I'll restore the original formatting

**Review by pcdavid — APPROVED** [review]

**Pull request merged**